### PR TITLE
Add/test site credentials

### PR DESCRIPTION
**Pull request by villanovachile**
#### Proposed Changes

* This PR seeks to fix the issue outlined in Asana task 1170120475965554-as-1202542076918872
* As explained in the Asana task, currently there is no check in place to test credentials for Jetpack connected site with Jetpack backup. WPCOM Calypso > Settings > Jetpack, as well as Cloud Jetpack > Settings shows "Connected" only if credentials have been entered. This can be problematic and provide an erroneous status if credentials have changed, causing confusion for both users and HEs alike.
* This PR makes use of a new endpoint in the public WPCOM API that will test the credentials and set the connected status accordingly

Some considerations: Every time a user navigates to Settings > Jetpack on WPCOM, or Settings on Cloud.jetpack.com, this will test the credentials. As such, I would not anticipate that this page is frequented unless a user is experiencing an issue with their credentials. Nevertheless, it's something I wanted to mention, in case this might be a concern.

#### Testing Instructions

* Spin up a new Jurassic Ninja site (or use an existing self hosted site), connect Jetpack, and add a Jetpack Backup plan
* Add server credentials at cloud.jetpack.com > Settings, or follow the steps here: https://jetpack.com/support/backup/backups-via-the-jetpack-plugin/adding-credentials-to-jetpack/
* Navigate to cloud.jetpack.com > Settings, or to WPCOM > Settings > Jetpack, and confirm that it shows connected (results should match test from RWDB)
* SSH into your site and change password, then navigate back to cloud.jetpack.com > Settings or to WPCOM > Settings > Jetpack, and Calypso should show Disconnected prompting user to re-enter credentials

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [X] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [X] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202542076918872

### CODE REVIEW

File: client/components/advanced-credentials/index.tsx
Comment (tyxla) [inline client/components/advanced-credentials/index.tsx]: These are usually worth to have some sort of error handling. We're already handling the success case, but what happens if the request fails?

File: client/components/advanced-credentials/index.tsx
Comment (tyxla) [inline client/components/advanced-credentials/index.tsx]: Should we put `testCredentialsLoading` right in the `isRequestingCredentials` selector function? Because right now someone could accidentally use `isRequestingCredentials` and miss `testCredentialsLoading`.

File: client/components/advanced-credentials/index.tsx
Comment (tyxla) [inline client/components/advanced-credentials/index.tsx]: Should we put `testCredentialsResult` in the `hasCredentials ` constant definition? Because right now someone could accidentally use `hasCredentials` and miss `testCredentialsResult`.

**Review by tyxla — COMMENTED** [review]
This generally looks alright from a frameworks perspective. 

Let's make sure all checks pass, though.

I've left a few comments with suggestions. Let me know if that helps, and if you need anything else.

I'd defer to your team for final review and approval, because I don't have access to Asana anymore and can't access the card/issue you've linked.

File: client/components/advanced-credentials/index.tsx
Comment (villanovachile) [reply client/components/advanced-credentials/index.tsx]: Thank you for the suggestion! @jwebbdev made #62607 earlier this year, which uses `hasCredentials` to test if a site is Atomic. I've put `testCredentialsResult` in the `hasCredentials` constant definition, and in my testing, I haven't seen a conflict with regards to Atomic detection. I wonder if for any reason `testCredentialsResult` returns false, it may affect Atomic detection. John, do you have any thoughts? I wonder if perhaps we should leave this line as is, or create a new const definition here.

Comment (jwebbdev) [reply client/components/advanced-credentials/index.tsx]: Yeah, the referenced isAtomic check wouldn't benefit from a connection test being run.  It is a more lightweight check to see the type of credentials that are saved.  

Because the credentials test isn't exactly light weight (it tests a connection to the customer's site over FTP/SSH) and currently doesn't have caching implemented we probably only want to call it explicitly.  Potentially a `hasValidCredentials` selector could do it, but we might want some caching in place before it's used too widely.

File: client/components/advanced-credentials/index.tsx
Comment (villanovachile) [reply client/components/advanced-credentials/index.tsx]: Thanks for the suggestions! Made changes as suggested!

File: client/components/advanced-credentials/index.tsx
Comment (villanovachile) [reply client/components/advanced-credentials/index.tsx]: Done, thanks for the suggestion! I think this does make it a bit cleaner and more efficient.

**Review by tyxla — COMMENTED** [review]
This looks great 👍  Nothing more to add from the code / frameworks perspective. 

As mentioned earlier, I'm leaving the final testing and approval to the Yamato Backup team.

**Comment by purplepioneer** [comment]
Just want to add that as a HE, if this change can be pushed through it would have a big impact on volume. 

At the moment, the majority of backup and restore failure tickets that come into the queue are due to us having invalid credentials on file. When we let users know they have invalid creds, there's at least a 50/50 chance the user will reply with a screenshot of Calypso's credentials test showing "connected," due to the cached result. 

If we're able to change things so that Calypso's credentials test reflects reality (if it could be linked to the credentials test in RWDB it would be ideal) rather than a cached "connected" result when we have invalid credentials on file, there will be a noticeable impact.

We're going to be able to tell users to update their credentials without receiving what is often an inevitable response back showing a screenshot of Calypso's credentials page showing "connected." This saves time for our users and will save a lot of time for HEs.

**Review by jwebbdev — APPROVED** [review]
Let's give this a shot @villanovachile .  We might want to use a similar test before we start a restore and the team can probably iterate on both if changes are needed.

**Comment by villanovachile** [comment]
> Let's give this a shot @villanovachile . We might want to use a similar test before we start a restore and the team can probably iterate on both if changes are needed.

Sounds like a plan, thanks again, John!

**Comment by fullofcaffeine** [comment]
Rebasing to see if E2Es pass.

**Comment by fullofcaffeine** [comment]
@villanovachile Not sure why GH only shows my avatar in the commits after I rebased and force-pushed. It used to show the avatar for both committers 😬 , but it does say you authored the commits. Feel free to rebase again and force-push if you'd like.

**Comment by villanovachile** [comment]
> @villanovachile Not sure why GH only shows my avatar in the commits after I rebased and force-pushed. It used to show the avatar for both committers 😬 , but it does say you authored the commits. Feel free to rebase again and force-push again if you'd like.

All good! I’ll deploy once the last test finishes. Thanks for the help, Marcelo!

**Comment by fullofcaffeine** [comment]
@villanovachile Looks like it only needed a rebase! All checks are passing now 🟢 🎉

**Pull request merged**
